src/coll/mod.rs

```diff
@@ -178,13 +180,31 @@ where
         self.inner.write_concern.as_ref()
     }
 
-    /// Drops the collection, deleting all data and indexes stored in it.
-    pub async fn drop(&self, options: impl Into<Option<DropCollectionOptions>>) -> Result<()> {
+    async fn drop_common(
+        &self,
+        options: impl Into<Option<DropCollectionOptions>>,
+        session: impl Into<Option<&mut ClientSession>>,
+    ) -> Result<()> {
         let mut options = options.into();
         resolve_options!(self, options, [write_concern]);
 
         let drop = DropCollection::new(self.namespace(), options);
-        self.client().execute_operation(drop).await
+        self.client().execute_operation(drop, session.into()).await
```

Comment:
My understanding was that we were planning on changing the return type of each of these methods so that that a user could either `await` and get back the result without a session (so that the API worked the same as before in that case) or could call a method like `with_session(&mut session)` and then `await` that to execute the operation with a session. Is there a reason that we didn't go that route with this implementation?

Reply:
The API here is in line with the proposed [design](https://docs.google.com/document/d/1xdIk3RVTUh_Wrna52nqChq1ZoqOBqbaak5ZvfwuhzEw/edit#heading=h.hjltafp465xm). Back when we were first looking at this, we didn't consider the builder approach too much since it would technically be a breaking change, and I don't think we really ever addressed it again. Now that we're doing 2.0, we could re-evaluate it, though all of this code has already been written.

Note that cursors do have a `with_session` in the design, since they can't retain ownership of the session or a reference to it without preventing other operations from using that session.

Reply:
Looking back at the [original-original design](https://gist.github.com/patrickfreed/e37d084500c0bdfe676556db1dc7cc1b#cons-2), it looks like a couple options were laid out without one explicitly being picked:
>This could entail duplicating the entire API or having users specify None whenever they don't need a session. Alternatively, we could use a builder pattern on the returned future to specify whether a session is required or not. e.g.
>```rust
>coll.find(doc! { "x": 1 }).with_session(&mut session).await; // Cursor<'a>
>coll.find(doc! { "x": 1 }).await; // Cursor<'static>
>coll.insert_one(doc! { "x": 1 }).with_session(&mut session).await;
>coll.insert_one(doc! { "x": 1 }).await;
>```
>This works nicely for async because we're always returning futures, but it doesn't apply directly to sync which has no intermediate return values. For sync, we'd either have to introduce them or just duplicate the API:
>```rust
>coll.find_with_session(&mut session, doc! { "x": 1 }); // Cursor<'a>
>coll.find(doc! { "x": 1 }); // Cursor<'static>
>coll.insert_one_with_session(&mut session, doc! { "x": 1 });
>coll.insert_one(doc! { "x": 1 }).await;
>```
It looks like consistency with the sync API was another concern, but again I think it mostly came down to not wanting to make a breaking change.

Reply:
Hm I wasn't aware of the alternative builder design; I used the first doc Patrick linked as a reference when implementing this. I personally find the current API a bit more straightforward, but do see the benefit in reducing duplication with the builder pattern. I'd be fine with re-implementing the other way if either of you feels strongly about it (although worth noting that a significant overhaul will probably delay us on being able to get transactions out a little longer).

Reply:
I don't feel strongly either way. Builder is slightly "prettier" maybe? but also slightly easier to forget to do and likely non-trivial to implement. I guess my opinion then would be to stick with the current implementation because it's already done and I personally don't see a huge difference between the two at the moment.

Reply:
I'm mostly not super thrilled with having to duplicate all the methods in our API, but if you both are fine with it, I'll resolve this thread.